src/cargo/ops/registry/publish.rs

```diff
@@ -57,187 +65,224 @@ pub struct PublishOpts<'gctx> {
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+            transmit(
+                opts.gctx,
+                ws,
+                pkg,
+                tarball.file(),
+                &mut registry,
+                source_ids.original,
+                opts.dry_run,
+            )?;
+
+            if !opts.dry_run {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
+            }
+        }
+
+        let finished = if opts.dry_run {
+            ready.clone()
         } else {
-            DEFAULT_TIMEOUT
+            const DEFAULT_TIMEOUT: u64 = 60;
+            let timeout = if opts.gctx.cli_unstable().publish_timeout {
+                let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
+                timeout.unwrap_or(DEFAULT_TIMEOUT)
+            } else {
+                DEFAULT_TIMEOUT
+            };
+            if 0 < timeout {
+                let timeout = Duration::from_secs(timeout);
+                wait_for_any_publish_confirmation(
+                    opts.gctx,
+                    source_ids.original,
+                    &to_confirm,
+                    timeout,
+                )?
+            } else {
+                Vec::new()
+            }
         };
-        if 0 < timeout {
-            let timeout = Duration::from_secs(timeout);
-            wait_for_publish(opts.gctx, source_ids.original, pkg, timeout)?;
+
+        for id in &finished {
+            to_confirm.remove(id);
+            remaining.remove(id);
+        }
+        if finished.is_empty() {
+            // If nothing finished, it means we timed out while waiting for confirmation.
+            // We're going to exit, but first we need to check: if we pretend that the unconfirmed
+            // uploads are confirmed, did we finish everything?
+            for id in &ready {
+                remaining.remove(id);
+            }
+
+            if remaining.is_empty() {
+                // It's ok that we timed out, because nothing was waiting on dependencies to
+                // be confirmed.
+                break;
+            } else {
+                let failed_list = package_list(remaining, "and");
+                bail!("failed to publish {failed_list} because we timed out waiting for dependencies.");
```

Comment:
Feels weird to say something failed that we haven't attempted
```suggestion
                bail!("unable to publish {failed_list} because we timed out waiting for dependencies.");
```